### PR TITLE
Reformat indentation on paste

### PR DESCRIPTION
**Pull request by saul**
paste

Notice how in the GIF indentation is automatically restructured depending on where you paste. The lack of this feature has been bothering me since day 1 as I cut/paste code around into different scopes often. Near enough feature complete already.

**Todo**:

- [x] Write tests
- [x] Fix indentation when next line should be indented
- [x] Add "Automatically format on paste" to the F# options

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs [hunk 1]
```diff
@@ -22,6 +23,10 @@ type internal FSharpEditorFormattingService
         checkerProvider: FSharpCheckerProvider,
         projectInfoManager: FSharpProjectOptionsManager
     ) =
+    
+    static let toIList (xs : 'a seq) : IList<'a> = ResizeArray(xs) :> _
```
Comment (TIHan) [inline vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs]: Just a nit: do ```static let toIList (xs : 'a seq) = ResizeArray(xs) :> IList<'a>```

**Comment by TIHan** [comment]
This is cool! It will make the experience much better. :)

File: vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs [hunk 2]
```diff
@@ -22,6 +23,10 @@ type internal FSharpEditorFormattingService
...
+    
+    static let getIndentation : string -> int = Seq.takeWhile ((=) ' ') >> Seq.length
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs]: And this is just after Don's talk where he said "do not use point free, just don't" ;)

File: vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs [hunk 3]
```diff
@@ -68,6 +73,79 @@ type internal FSharpEditorFormattingService
...
+                        None
+                )
+            
+            if leadingIndentation > 0 then
+                return stripIndentation leadingIndentation
+            else
+                let nextLineShouldBeIndented = FSharpIndentationService.IndentShouldFollow(documentId, sourceText, filePath, span.Start, parsingOptions)
+                
+                let removeIndentation =
+                    let nextLineIndent = fixedPasteText.Lines.[1].ToString() |> getIndentation
```
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs]: Is `.[1]` is safe here?

Comment (saul) [reply vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs]: There’s a check above that checks that the span is multi line

**Comment by vasily-kirichenko** [comment]
It does not work for me:

untitled

**Comment by saul** [comment]
Do you have indentation mode set to 'Smart'?

**Comment by vasily-kirichenko** [comment]
Yes.

**Comment by vasily-kirichenko** [comment]
I've restarted VS, the same behaviour.

**Comment by vasily-kirichenko** [comment]
The vsix seems to be installed:

image

**Comment by vasily-kirichenko** [comment]
image

**Comment by saul** [comment]
paste2

Your code works on my machine 🤷‍♂️

**Comment by vasily-kirichenko** [comment]
But the code looks like this at one moment:

image

Is it ok?..

**Comment by vasily-kirichenko** [comment]
Sorry, maybe I don't understand how the feature should work...

**Comment by saul** [comment]
Well that's where my caret was when I hit paste. I did have to manually de-indent to that level... I was copying your GIF verbatim.

**Comment by vasily-kirichenko** [comment]
I'm lost :( So, everything is OK?

**Comment by nosami** [comment]
This looks to be how it works in VS for Mac too. Indentation of the paste is adjusted to the caret column. Was one of the first things I worked on.

**Comment by saul** [comment]
@vasily-kirichenko maybe this GIF demonstrates what I'm trying to say better. Notice where the caret is when I hit paste - that's the only thing that matters:

paste3

**Comment by nosami** [comment]
I can't tell from the gif, but does it also work when the selection is irregular like this (the selection doesn't start from column 0, but subsequent lines do) ?

indent

**Comment by saul** [comment]
Indeed it does - see the GIF in the PR description.

**Comment by vasily-kirichenko** [comment]
@saul Ah, I see. Thanks :)

File: vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs (hunk 3 above)
Comment (TIHan) [reply vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs]: It seems difficult to see that check. Can we explicitly guard right before it for sanity?

**Comment by cartermp** [comment]
Tet failure is in Async and seems unrelated to this PR. @dsyme?

**Comment by saul** [comment]
Unit tests are failing to run in part1, but Jenkins is still passing:

```
---------------- Done with update, starting tests -----------------------
WHERE_ARG_NUNIT=
FSCBINPATH=D:\j\w\release_ci_pa---3f142ccc\release\net40\bin
RESULTSDIR=D:\j\w\release_ci_pa---3f142ccc\tests\TestResults
link_exe=D:\j\w\release_ci_pa---3f142ccc\tests\fsharpqa\testenv\bin\link\link.exe
NUNIT3_CONSOLE=D:\j\w\release_ci_pa---3f142ccc\packages\NUnit.Console.3.0.0\tools\nunit3-console.exe
NUNITPATH=D:\j\w\release_ci_pa---3f142ccc\tests\fsharpqa\testenv\bin\nunit\
"D:\j\w\release_ci_pa---3f142ccc\packages\NUnit.Console.3.0.0\tools\nunit3-console.exe" --verbose --x86 --framework:V4.0 --result:"D:\j\w\release_ci_pa---3f142ccc\tests\TestResults\test-vs-ideunit-results.xml;format=nunit3" --output:"D:\j\w\release_ci_pa---3f142ccc\tests\TestResults\test-vs-ideunit-output.log"  --err:"D:\j\w\release_ci_pa---3f142ccc\tests\TestResults\test-vs-ideunit-errors.log"  --work:"D:\j\w\release_ci_pa---3f142ccc\release\net40\bin"  --workers=1 --agents=1 --full "D:\j\w\release_ci_pa---3f142ccc\release\net40\bin\VisualFSharp.UnitTests.dll" 
NUnit Console Runner 3.0.5797 
Copyright (C) 2015 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.14393.0
  CLR Version: 4.0.30319.42000

Test Files
    D:\j\w\release_ci_pa---3f142ccc\release\net40\bin\VisualFSharp.UnitTests.dll

Agent process 2432 starting
Running under version 4.0.30319.42000, Net 4.5
Agent process 2432 exiting

Errors and Failures

1) Invalid : D:\j\w\release_ci_pa---3f142ccc\release\net40\bin\VisualFSharp.UnitTests.dll
Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.

Run Settings
    RuntimeFramework: V4.0
    RunAsX86: True
    WorkDirectory: D:\j\w\release_ci_pa---3f142ccc\release\net40\bin
    MaxAgents: 1
    NumberOfTestWorkers: 1
    Verbose: True

Test Run Summary
    Overall result: Failed
   Tests run: 0, Passed: 0, Errors: 0, Failures: 0, Inconclusive: 0
     Not run: 0, Invalid: 0, Ignored: 0, Explicit: 0, Skipped: 0
  Start time: 2018-05-16 13:37:27Z
    End time: 2018-05-16 13:37:27Z
    Duration: 0.069 seconds

Results (nunit3) saved as D:\j\w\release_ci_pa---3f142ccc\tests\TestResults\test-vs-ideunit-results.xml
[Current build status] check if current [SUCCESS] is worse or equals then [FAILURE] and better or equals then [SUCCESS]
Run condition [Current build status] enabling perform for step [[Archive the artifacts]]
Archiving artifacts
[Current build status] check if current [SUCCESS] is worse or equals then [FAILURE] and better or equals then [SUCCESS]
Run condition [Current build status] enabling perform for step [[Archive the artifacts]]
Archiving artifacts
Notifying endpoint with credentials id 'helix-int-notification-url'
Notifying endpoint with credentials id 'helix-prod-notification-url'
Setting status of 6ff6749327a11bce405731414d576b849b4c34ab to SUCCESS with url https://ci2.dot.net/job/Microsoft_visualfsharp/job/master/job/release_ci_part1_windows_nt_prtest/6235/ and message: 'Build finished. '
Using context: Windows_NT Release_ci_part1 Build
[WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
Finished: SUCCESS
```

**Comment by saul** [comment]
Once I can get the unit tests to run, I can fix them up (if they're not green). I'm getting the same "Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information." on my machine as Jenkins is - I'm not sure how I could have caused that.

Any ideas?

Also @TIHan are you able to review this?

**Comment by cartermp** [comment]
@saul I believe this is related to #4920, which is one of the root causes of #4850, which is affecting you here. It's a bit of a fire drill right now 😱

**Comment by saul** [comment]
@KevinRansom I've merged from master and stuff the same issue.

**Comment by saul** [comment]
@cartermp any update on how I can get these tests running again?

**Comment by cartermp** [comment]
@KevinRansom is still on point to clean out the bit rot in some legacy tests. Until then, no updates.

**Comment by saul** [comment]
But a load of otherwise functioning tests have been disabled?

**Comment by vasily-kirichenko** [comment]
I find it absolutely unacceptable to happily merge PRs having _a lot_ of tests disabled. 

/cc @dsyme

**Comment by saul** [comment]
This PR is also blocked on the fact that I want to write unit tests for new functionality but can't actually test it 😢

**Comment by vasily-kirichenko** [comment]
@saul it's even more strange that some tests from Roslyn folder are passed locally (e.g. ColorizationServiceTests), but are not enabled on CI.

**Comment by cartermp** [comment]
@saul 

> But a load of otherwise functioning tests have been disabled?

I'll have to defer to @KevinRansom on the best way to proceed w.r.t tests, as this isn't my call. Most tests should be just fine, and I believe the Roslyn-based ones have no issues with bit-rot at all, since they're the ones which have been actively developed.

**Comment by KevinRansom** [comment]
@saul -- I'm sorry you are blocked.  I had to switch away to 4.5 for a short while.

**Comment by saul** [comment]
I’m on vacation for 10 days - I’ll take a look at this when I get back. I really want this in!

**Comment by cartermp** [comment]
Cheers, and happy vacation!


On Sat, Jul 21, 2018, at 13:53, Saul Rennison wrote:
> I’m on vacation for 10 days - I’ll take a look at this when I get
> back. I really want this in!> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[1], or mute the thread[2].

Links:

  1. https://github.com/Microsoft/visualfsharp/pull/4702#issuecomment-406822926

**Comment by saul** [comment]
@dotnet-bot test this please

**Comment by cartermp** [comment]
@saul We'd love to take this for VS 2019 Previews. Do you want to resolve conflicts?

**Comment by saul** [comment]
This is pretty much finished - I’ll try running (and fixing) the tests when I’m back from NY.

**Comment by cartermp** [comment]
You should _never_ stop going on vacation (especially if you ever end up working in the US, where vacation is a rarity)! No pressure at all.

**Comment by saul** [comment]
Tests are green, this is ready to merge @cartermp.

BTW - haven't really done any work in this repo for a few months, but the Azure DevOps is way nicer than what was there before. Also being able to run unit tests in the IDE (instead of them being EXEs) made debugging this significantly easier. A lot of contributor pain has gone since I last worked on this repo! Great work guys.

**Comment by KevinRansom** [comment]
@cartermp,

I will go through the resources files and fix them up.

**Comment by cartermp** [comment]
Spoke with @KevinRansom and these files are build outputs that are an unfortunate consequence of our build today. This issue shouldn't persist once the .NET SDK work is completed.

**Pull request merged**